rasgoql/rasgoql/data/postgres.py

```diff
@@ -0,0 +1,630 @@
+        """
+        Returns a dict of the credentials
+        """
+        return {
+            "dialect": self.dialect,
+            "db_api": self.db_api,
+            "username": self.username,
+            "password": self.password,
+            "host": self.host,
+            "port": self.port,
+            "default_db": self.default_db,
+            "default_schema": self.default_schema
+        }
+
+    def to_env(
+            self,
+            filepath: str = None,
+            overwrite: bool = False
+        ):
+        """
+        Saves credentials to a .env file on your machine
+        """
+        creds = {
+            "POSTGRES_DIALECT": self.dialect,
+            "POSTGRES_DB_API": self.db_api,
+            "POSTGRES_USERNAME": self.username,
+            "POSTGRES_PASSWORD": self.password,
+            "POSTGRES_HOST": self.host,
+            "POSTGRES_PORT": self.port,
+            "POSTGRES_DEFAULT_DB": self.default_db,
+            "POSTGRES_DEFAULT_SCHEMA": self.default_schema
+        }
+        return save_env(creds, filepath, overwrite)
+
+
+class PostgresDataWarehouse(DataWarehouse):
+    """
+    Postgres DataWarehouse
+    """
+    dw_type = 'postgres'
+    credentials_class = PostgresCredentials
+
+    def __init__(self):
+        super().__init__()
+        self.credentials: dict = None
+        self.connection: alchemy_engine = None
+        self.default_database = None
+        self.default_schema = None
+
+    # ---------------------------
+    # Core Data Warehouse methods
+    # ---------------------------
+    def change_namespace(
+            self,
+            namespace: str
+        ) -> None:
+        """
+        Changes the default namespace of your connection
+
+        Params:
+        `namespace`: str:
+            namespace (database.schema)
+        """
+        # TODO: changing databases is unsupported in Postgres - we would need to rebuild a new connection. Error here?
+        raise NotImplementedError("Connecting to a new Database in a single session is not supported by Postgres. Please build a new connection using the PostgresCredentials class")
+
+    def connect(
+            self,
+            credentials: Union[dict, PostgresCredentials]
+        ):
+        """
+        Connect to Postgres
+
+        Params:
+        `credentials`: dict:
+            dict (or DWCredentials class) holding the connection credentials
+        """
+        if isinstance(credentials, PostgresCredentials):
+            credentials = credentials.to_dict()
+
+        try:
+            self.credentials = credentials
+            self.default_database = credentials.get('default_db')
+            self.default_schema = credentials.get('default_schema')
+            self.engine_url = f"{credentials.get('dialect')}+{credentials.get('db_api')}://{credentials.get('username')}:{credentials.get('password')}@{credentials.get('host')}:{credentials.get('port')}/{credentials.get('default_db')}"
```

Comment:
Debating whether there's more value in abstracting some of these concepts to match other dbs or exposing them. Let's talk about it